citrination_client/search/client.py

```diff
@@ -0,0 +1,200 @@
+from .query_encoder import QueryEncoder
+from pypif.util.case import to_camel_case
+from pypif.util.case import keys_to_snake_case
+from citrination_client.search import *
+from citrination_client.util import config as client_config
+from citrination_client.base.base_client import BaseClient
+from citrination_client.util import http as http_util
+
+from copy import deepcopy
+from time import sleep
+import json
+import requests
+import routes
+
+class SearchClient(BaseClient):
+
+    def __init__(self, api_key, webserver_host="https://citrination.com"):
+        members = [
+            "pif_search",
+            "pif_multi_search",
+            "dataset_search"
+        ]
+        super(SearchClient, self).__init__(api_key, webserver_host, members)
+
+    def pif_search(self, pif_system_returning_query):
+        """
+        Run a PIF query against Citrination.
+
+        :param pif_system_returning_query: :class:`PifSystemReturningQuery` to execute.
+        :return: :class:`PifSearchResult` object with the results of the query.
+        """
+        return self._execute_paginating_search(
+            pif_system_returning_query,
+            PifSearchResult
+        )
+
+    def dataset_search(self, dataset_returning_query):
+        """
+        Run a dataset query against Citrination.
+
+        :param dataset_returning_query: :class:`DatasetReturningQuery` to execute.
+        :return: :class:`DatasetSearchResult` object with the results of the query.
+        """
+        return self._execute_paginating_search(
+            dataset_returning_query,
+            DatasetSearchResult
+        )
+
+    def _execute_paginating_search(self, returning_query, result_class):
+        """
+        Run a PIF query against Citrination.
+
+        :param returning_query: :class:`BaseReturningQuery` to execute.
+        :param result_class: The class of the result to return.
+        :return: ``result_class`` object with the results of the query.
+        """
+        if returning_query.from_index:
+            from_index = returning_query.from_index
+        else:
+            from_index = 0
+
+        if returning_query.size != None:
+            size = min(returning_query.size, client_config.max_query_size)
+        else:
+            size = client_config.max_query_size
+
+        if (size == client_config.max_query_size and
+            size != returning_query.size):
+            self._warn("Query size greater than max system size - only {} results will be returned".format(size))
+
+        time = 0.0; hits = []; first = True
+        while True:
+            if first:
+                first = False
+            else:
+                sleep(1)
```

Comment:
check if enough for rate limiting

Reply:
check for rate limiting exception in the BaseClient and look for the rate limiting information and retry there

Reply:
👍